Design/ssa-conversion.md

```diff
@@ -0,0 +1,101 @@
+# Converting an ExpEmbedding into SSA form
+
+In order to translate pure Kotlin functions into Viper functions their body must
+be translated into a single expression. Therefore, reassignments of variables may
+not happen on the Viper level which is why the ExpEmbedding must be translated
+into SSA form.
+
+## Converting a Block to SSA form
+
+Assuming we have a block embedding with no branching (branching in the CFG that
+is), converting into SSA form is quite straight-forward:
+- Traverse the ExpressionEmbeddings in the block in order
+- As soon as a reassignment is found introduce a new variable and assign it the
+  RHS of the reassignment.
+- In subsequent usages of the 'original' variable use the newly introduced
+  version.
+In code this may look like the following:
+```kotlin
+var x = 1
+x = x + 1
+return x
+```
+translates to:
+```viper
+let x_0 == (1) in
+let x_1 == (x_0 + 1) in
+x_1
+```
+Note: An optimized encoding of the above example could be created by assigning
+1 + 1 directly to x_0. In general, if no usages occur between two reassignments
+of the same variable we can omit one assignment.
```

Comment:
This is mixing up two different things:
- If an assigned value is never read, you can drop the assignment.
- If a variable is only used once, then you can inline its definition.

In SSA form, you see this as zero or one usage of a variable respectively.  However, I would not worry about it for now.